cdap-hbase-compat-0.98/src/main/java/co/cask/cdap/data2/transaction/messaging/coprocessor/hbase98/MessageTableRegionObserver.java

```diff
@@ -89,63 +89,84 @@
       HTableDescriptor tableDesc = ((RegionCoprocessorEnvironment) env).getRegion().getTableDesc();
-      String metadataTableNamespace = tableDesc.getValue(Constants.MessagingSystem.HBASE_METADATA_TABLE_NAMESPACE);
-      String hbaseNamespacePrefix = tableDesc.getValue(Constants.Dataset.TABLE_PREFIX);
+      metadataTableNamespace = tableDesc.getValue(Constants.MessagingSystem.HBASE_METADATA_TABLE_NAMESPACE);
+      hbaseNamespacePrefix = tableDesc.getValue(Constants.Dataset.TABLE_PREFIX);
+      prefixLength = Integer.valueOf(tableDesc.getValue(
+        Constants.MessagingSystem.HBASE_MESSAGING_TABLE_PREFIX_NUM_BYTES));
 
       HTableNameConverter nameConverter = new HTable98NameConverter();
       String sysConfigTablePrefix = nameConverter.getSysConfigTablePrefix(hbaseNamespacePrefix);
       cConfReader = new CConfigurationReader(env.getConfiguration(), sysConfigTablePrefix);
-      CConfiguration cConf = cConfReader.read();
-      String metadataTableName = cConf.get(Constants.MessagingSystem.METADATA_TABLE_NAME);
-
-      prefixLength = Integer.valueOf(tableDesc.getValue(
-        Constants.MessagingSystem.HBASE_MESSAGING_TABLE_PREFIX_NUM_BYTES));
-
-      metadataTable = env.getTable(nameConverter.toTableName(hbaseNamespacePrefix,
-                                                             TableId.from(metadataTableNamespace, metadataTableName)));
-      final long metadataCacheExpiry = cConf.getLong(
-        Constants.MessagingSystem.COPROCESSOR_METADATA_CACHE_EXPIRATION_SECONDS);
-      topicCache = CacheBuilder.newBuilder()
-        .expireAfterWrite(metadataCacheExpiry, TimeUnit.SECONDS)
-        .maximumSize(1000)
-        .build(new CacheLoader<ByteBuffer, Map<String, String>>() {
-
-          @Override
-          public Map<String, String> load(ByteBuffer topicBytes) throws Exception {
-            byte[] getBytes = topicBytes.array().length == topicBytes.remaining() ?
-              topicBytes.array() : Bytes.toBytes(topicBytes);
-            Get get = new Get(getBytes);
-            Result result = metadataTable.get(get);
-            byte[] properties = result.getValue(COL_FAMILY, COL);
-            Map<String, String> propertyMap = GSON.fromJson(Bytes.toString(properties), MAP_TYPE);
-            String ttl = propertyMap.get(MessagingUtils.Constants.TTL_KEY);
-            long ttlInMs = TimeUnit.SECONDS.toMillis(Long.parseLong(ttl));
-            propertyMap.put(MessagingUtils.Constants.TTL_KEY, Long.toString(ttlInMs));
-            return propertyMap;
-          }
-        });
 
       Supplier<TransactionStateCache> cacheSupplier = getTransactionStateCacheSupplier(hbaseNamespacePrefix,
                                                                                        env.getConfiguration());
       txStateCache = cacheSupplier.get();
     }
   }
 
+  private synchronized LoadingCache<ByteBuffer, Map<String, String>> getTopicCache(RegionCoprocessorEnvironment env)
```

Comment:
Yes, it needs to be synchronized, since we want to initialize topicCache only once (on demand).